### ISSUE TITLE
Script integrity should not score higher than prohibiting scripts at all

### ISSUE THREAD

**Issue by 5ecede** [mozilla/http-observatory#273]
Websites without any (Java)script avoid an important attack vector, but as JS is heavily used, it's great that there is a way to stop browsers from executing modified scripts. The Observatory rightfully gives bonus points to websites that include the script integrity hash value.

It does not seem possible to hash the nonexistance of scripts, or to generate an "empty whitelist". (If there is, please tell me how!) In my opinion the avoidance of JS (with a csp of "script-sri 'none'" and, in the future,  "require-sri-for: script") could perhaps considered to be an equivalent achievement in making the web a safer place.

**Comment by floatingatoll** [comment]
Nonexistence of scripts is enforced through the CSP header, at which point
SRI no longer applies:

Content-Security-Policy: script-src: 'none';

That generates the 'empty whitelist' you seek in practice.

**Comment by floatingatoll** [comment]
(Sorry, hit send too soon.)

Are you proposing that CSP: script-src: 'none' should receive a +5 bonus equal to that received by SRI?

Sites that import external JS resources without SRI are vulnerable to "attacks at a distance", where a third-party .JS file is altered unknowning to all users of it. SRI protects against this specific attack, and the Observatory offers bonuses for SRI as a result.

Both sites that do and do not import external JS resources are vulnerable to "page modification", where either an extension or a proxy modifies the page content to include an unexpected resource. Sites are actively penalized by the Observatory for failing to restrict the sources from which scripts are loaded, and that penalty is removed when they implement a script-src or equivalent directive.

The score 'boost' provided by removing the 'no script-src' penalty is hidden from your view, while the score boost provided by going the extra mile with SRI is not. If you look at the detailed scoring options for sites, I believe there's at *least* a +5 difference solely due to having 'script-src' specified, and moreso when it contains safe values.

This does mean that sites without JavaScript and CSS will have a potential maximum score 5 points less than sites that do have either or both of those SRI-capable resources, but I personally think that's fine. Most of our year here in IT has been spent getting sites up to 90/100, which is a significant amount of work; I look forward someday to finding out that we're capped at 140/100 on a site when we could get 145, simply because we specified 'script-src: none'. It'll be slightly ironic, but by then, we'll have secured that site the best of all them, and will be proud to have reached the end of the scoring 'game'.

**Comment by 5ecede** [comment]
Yes, that was my proposal. I noticed that the current maximum value of 135 cannot be achieved by websites avoiding Javascript. CSS-SRI can't provide the 5 points in question, they are a separate issue. 

It's not a big problem for me, I used the weekend to continue to teach myself on CSP and security headers, and when I tried to implement them I _only_ got to a limit of 125, not 130 (excluding the points available for HPKP for obvious reasons).

Yes, script-src and especially default 'none' really help to gain quite a few points. 

Thank you for your answer. I can absolutely live with this situation, and I appreciate the Observatory, as it was really helpful on my path to secure my humble test website. :)

**Comment by april** [comment]
This is actually something I considered when I first implemented the SRI test. My problem is that awarding +5 for not having scripts seems really weird. Since it's mostly just for bragging rights, I figured it wasn't that big of a deal. Maybe I'll change the presentation or something on the website to make it appear as 135 even if you only got 130, for this particular case.

**Comment by floatingatoll** [comment]
With the advent of caching layers hosted by third parties, I could also
make a case that we should drop the SRI scores by 5 across the board, so
that -10 for offsite-no-sri, -5 for onsite-no-sri, 0 for sri-or-no-scripts.
